core/trino-main/src/main/java/io/trino/sql/planner/optimizations/ArrayAggregationAfterDistinctOptimizer.java

```diff
@@ -0,0 +1,106 @@
+import io.trino.Session;
+import io.trino.execution.warnings.WarningCollector;
+import io.trino.metadata.Metadata;
+import io.trino.operator.aggregation.arrayagg.ArrayAggregationFunction;
+import io.trino.operator.scalar.ArrayDistinctFunction;
+import io.trino.sql.planner.PlanNodeIdAllocator;
+import io.trino.sql.planner.Symbol;
+import io.trino.sql.planner.SymbolAllocator;
+import io.trino.sql.planner.TypeProvider;
+import io.trino.sql.planner.plan.AggregationNode;
+import io.trino.sql.planner.plan.Assignments;
+import io.trino.sql.planner.plan.PlanNode;
+import io.trino.sql.planner.plan.ProjectNode;
+import io.trino.sql.planner.plan.SimplePlanRewriter;
+import io.trino.sql.tree.FunctionCall;
+
+import static java.util.Objects.requireNonNull;
+
+public class ArrayAggregationAfterDistinctOptimizer
+        implements PlanOptimizer
+{
+    private final Metadata metadata;
+
+    public ArrayAggregationAfterDistinctOptimizer(Metadata metadata)
+    {
+        this.metadata = metadata;
+    }
+
+    @Override
+    public PlanNode optimize(PlanNode plan, Session session, TypeProvider types, SymbolAllocator symbolAllocator, PlanNodeIdAllocator idAllocator, WarningCollector warningCollector)
+    {
+        return SimplePlanRewriter.rewriteWith(new ArrayAggregationAfterDistinctOptimizer.Optimizer(metadata, idAllocator), plan);
+    }
+
+    private static class Optimizer
+            extends SimplePlanRewriter<Void>
+    {
+        private final Metadata metadata;
+        private final PlanNodeIdAllocator idAllocator;
+
+        private Optimizer(Metadata metadata, PlanNodeIdAllocator idAllocator)
+        {
+            this.metadata = metadata;
+            this.idAllocator = requireNonNull(idAllocator, "idAllocator is null");
+        }
+
+        @Override
+        public PlanNode visitProject(ProjectNode node, RewriteContext<Void> context)
+        {
+            PlanNode rewritten = context.defaultRewrite(node);
+            if (rewritten instanceof ProjectNode &&
+                    ((ProjectNode) rewritten).getSource() instanceof AggregationNode &&
+                    ((ProjectNode) rewritten).getAssignments().getExpressions().stream()
+                            .anyMatch(f -> f instanceof FunctionCall &&
+                                    metadata.decodeFunction(((FunctionCall) f).getName()).getSignature().getName().equals(ArrayDistinctFunction.NAME)) &&
+                    ((AggregationNode) ((ProjectNode) rewritten).getSource()).getAggregations().values().stream()
+                            .anyMatch(f -> f.getResolvedFunction().getSignature().getName().equals(ArrayAggregationFunction.NAME))) {
+                ProjectNode rewrittenProjectNode = (ProjectNode) rewritten;
+                AggregationNode source = (AggregationNode) rewrittenProjectNode.getSource();
+                AggregationNode.Aggregation aggregation = source.getAggregations().get(new Symbol(ArrayAggregationFunction.NAME));
```

Comment:
Do not match on the Symbol name.
Symbols might be replaced (renamed) during optimization. Also, if you have several calls to `array_agg` in the plan, they will be initially planned with different names for disambiguation.